include/gzip/compress.hpp

```diff
@@ -10,103 +10,100 @@
+    //  8 to 15 for zlib
+    // (8 to 15) + 16 for gzip
+    // (8 to 15) + 32 to automatically detect gzip/zlib header (decompression/inflate only)
+    constexpr int window_bits = 15 + 16; // gzip with windowbits of 15
+
+    constexpr int mem_level = 8;
+    // The memory requirements for deflate are (in bytes):
+    // (1 << (window_bits+2)) +  (1 << (mem_level+9))
+    // with a default value of 8 for mem_level and our window_bits of 15
+    // this is 128Kb
 
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wold-style-cast"
-        if (deflateInit2(&deflate_s, level_, Z_DEFLATED, window_bits, mem_level, Z_DEFAULT_STRATEGY) != Z_OK)
-        {
-            throw std::runtime_error("deflate init failed");
-        }
+    if (deflateInit2(&deflate_s, level, Z_DEFLATED, window_bits, mem_level, Z_DEFAULT_STRATEGY) != Z_OK)
+    {
+        throw std::runtime_error("deflate init failed");
+    }
 #pragma GCC diagnostic pop
 
+    std::string buffer;
+    do
+    {
+        constexpr std::size_t max_step = static_cast<std::size_t>(std::numeric_limits<unsigned int>::max());
+        unsigned int step_size = size > max_step ? max_step : static_cast<unsigned int>(size);
+        size -= step_size;
+        unsigned int buffer_size = buffering_size > step_size ? step_size : static_cast<unsigned int>(buffering_size);
+
         deflate_s.next_in = reinterpret_cast<z_const Bytef*>(data);
-        deflate_s.avail_in = static_cast<unsigned int>(size);
+        data = data + step_size;
+        deflate_s.avail_in = step_size;
 
-        std::size_t size_compressed = 0;
+        buffer.resize(static_cast<std::size_t>(buffer_size));
         do
         {
-            size_t increase = size / 2 + 1024;
-            if (output.size() < (size_compressed + increase))
-            {
-                output.resize(size_compressed + increase);
-            }
-            // There is no way we see that "increase" would not fit in an unsigned int,
-            // hence we use static cast here to avoid -Wshorten-64-to-32 error
-            deflate_s.avail_out = static_cast<unsigned int>(increase);
-            deflate_s.next_out = reinterpret_cast<Bytef*>((&output[0] + size_compressed));
+            deflate_s.avail_out = buffer_size;
+            deflate_s.next_out = reinterpret_cast<Bytef*>(&buffer[0]);
             // From http://www.zlib.net/zlib_how.html
             // "deflate() has a return value that can indicate errors, yet we do not check it here.
             // Why not? Well, it turns out that deflate() can do no wrong here."
             // Basically only possible error is from deflateInit not working properly
             deflate(&deflate_s, Z_FINISH);
-            size_compressed += (increase - deflate_s.avail_out);
+            output.append(buffer, 0, static_cast<std::size_t>(buffer_size - deflate_s.avail_out));
         } while (deflate_s.avail_out == 0);
+    } while (size > 0);
+    deflateEnd(&deflate_s);
```

Comment:
The only places I could see an exception being thrown would be from `std::string::resize` or `std::string::append`.  Do you think its worth wrapping it all in a large try catch?

Reply:
You could argue that you will only get exceptions if you run out of memory, in which case you are screwed anyways. It would be cleaner to do it correctly, but at least it should be documented, so the next person to change the code knows that this isn't an oversight but a design decision.